### PR TITLE
colexec: partially fix memory accounting in the external sorter

### PR DESCRIPTION
**Pull request by yuzefovich**
**colexec: speed up external operator tests**

We recently merged several changes to improve the handling of memory
limits, including by the external operators. However, this caused those
operators to use very low (1 byte) memory limits which can result in
some tests taking really long time. This commit updates the memory limit
in "force disk spill" scenarios to be the default one to speed the tests
up (this is acceptable because the limit only matters after the spilling
has already occurred).

Additionally, this commit fixes the test which exercises the fallback to
sort + merge join in the external hash joiner so that the fallback
always occurs (via forced repartitioning).

Release note: None

**colexec: add simple test of memory accounting in external sort**

This commit adds a sanity check for the memory accounting done
throughout the external sort operation. At the moment there
are a lot of known problems with the memory accounting, so the
test is not very strict. The goal of the test is to make sure
that the total maximum reported memory usage (as would have
been collected as stats) is within reasonable range. It additionally
checks that the number of partitions created is as expected too.

Release note: None

**coldata: fix ProportionalSize calculation for windows into Bytes vectors**

Previously, we were incorrectly computing `ProportionalSize` for Bytes
vectors that are actually "windows". In such cases, it is likely that
the start of the relevant chunk of data is not from the beginning of
`data` slice, but rather from the first offset value. This resulted in
incorrectly determining the size of a partition for the external sort.

The impact of this mistake is that we would report higher memory usage
by the external sorter than it actually is since all of the memory usage
registered with the standalone memory account is actually
double-counting of the memory used by the in-memory sort operator that
has just spilled because all batches are windows into the buffered batch
(i.e. we don't allocate any new memory). Therefore, the more precise
accounting we have, the lower double-counting is.

Another issue is that we would make partitions of smaller size than we
could.

Release note: None

**colexec: remove double-counting of some memory in external sort**

Previously, we would be registering the memory usage of the partition
that is currently being spilled twice:
- in a standalone memory account in `inputPartitioningOperator`, this is
needed to know when to start a new partition
- in the in-memory sorter used by the external sort to sort each
partition, this would happen during "spooling" phase.

However, the first usage is "fake" - we don't actually buffer all of the
data flowing through the input partitioning operator, so we would
effectively double-count some memory. This commit fixes that behavior by
removing the memory account altogether because a simple variable is
sufficient to determine where the partitions' boundaries are.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
It is extremely difficult to come up with expectations around the memory accounting in the external sorter (there are just so many components involved, we also randomize `coldata.BatchSize()`, the behavior of `append` is not easy to predict (we look at the capacities of `[]byte` when calculating footprint)).

Anyway, I added a unit test that hopefully will not be too flakey and will be improved over time. It does already show the impact of fixes, and I believe it would be worth it to backport the second and the third commits. Let's discuss.

**Comment by asubiotto** [comment]
OK. I think the key is to discuss the impact/risk tradeoff but I'm willing to be convinced.

**Review by asubiotto — APPROVED** [review]
:lgtm: 

> Reviewed 1 of 1 files at r1, 5 of 5 files at r2, 3 of 3 files at r3, 6 of 6 files at r4, 5 of 5 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60284#-:-MTFZ8dW2AP524J7WIS4:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Comment by yuzefovich** [comment]
TFTR!

bors r+

**Comment by yuzefovich** [comment]
Race tests timed out for `colexec` package when merging this PR caused by changed in #59851, so I added another commit that speeds the tests up (https://github.com/cockroachdb/cockroach/pull/60284/commits/629a5f169f7acd6e747ea4141b55ccb8763f4946). I believe the commit is very safe, so I'll go ahead and merge without waiting for approval for it.

bors r+

**Pull request merged**
